### PR TITLE
feat: add a host-based auth cache as a fallback

### LINKED ISSUE #650: Prevent extra authentication round trips in non-compliant registries
I have been testing with JFrog Artifactory and `oras cp` will receive a 401 for each and every access to the repo that it has to go resolve by following the auth challenge and then use the received token (actually it is the same as stored in the auth cache already but under a different scoped key) to make the request and succeed.  While the current behavior works, it is obviously inefficient.  `crane` and `docker` use a simplified host-based caching (not scope based caching like oras) so they do not have an issue in this specific case.

The discussion over on [slack](https://cloud-native.slack.com/archives/CJ1KHJM5Z/p1699538327616879) has lots of details.

### PR DESCRIPTION
**Pull request by ktarplee**
This PR adds a new exported method called `NewRobustCache()` and changes the `DefaultCache` to use the robust cache.  The robust cache uses scoped based auth but falls back to host based auth to better handle the situations described in #650 but retain the benefits of scoped based auth (where the token for a repo might be different than another repo in the same registry).

Closes #650

### CODE REVIEW

**Comment by ktarplee** [comment]
This PR obviously still needs 

- [ ] Unit tests
- [ ] Integration testing
- [ ] Documentation

File: registry/remote/auth/cache.go [hunk 1]
```diff
@@ -157,3 +157,67 @@ func (noCache) GetToken(ctx context.Context, registry string, scheme Scheme, key
...
+	}
+
+	// fallback
+	return fc.secondary.GetScheme(ctx, registry)
+}
+
+// GetToken implements Cache.
+func (fc *fallbackCache) GetToken(ctx context.Context, registry string, scheme Scheme, key string) (string, error) {
+	token, err := fc.primary.GetToken(ctx, registry, scheme, key)
+	if err == nil {
```
Comment (ktarplee) [inline registry/remote/auth/cache.go:196]: Not sure if we want to check for `errdef.ErrNotFound` and return other errors.

**Comment by shizhMSFT** [comment]
@ktarplee Would you like to update this PR to reflect the latest discussion in #650?

**Comment by ktarplee** [comment]
> @ktarplee Would you like to update this PR to reflect the latest discussion in #650?

I made the name change.

**Comment by ktarplee** [comment]
I also added unit tests for the hostCache and exported `NewSingleContextCache()` function. @shizhMSFT

File: registry/remote/auth/cache.go [hunk 2]
```diff
@@ -157,3 +157,76 @@ func (noCache) GetToken(ctx context.Context, registry string, scheme Scheme, key
...
+
+	// fallback
+	return fc.secondary.GetToken(ctx, registry, scheme, key)
+}
+
+// Set implements Cache.
+func (fc *fallbackCache) Set(ctx context.Context, registry string, scheme Scheme, key string, fetch func(context.Context) (string, error)) (string, error) {
+	token, err := fc.primary.Set(ctx, registry, scheme, key, fetch)
+	if err != nil {
+		return token, err
```
Comment (shizhMSFT) [inline registry/remote/auth/cache.go:208]: nit:
```suggestion
		return "", err
```

**Review by shizhMSFT — APPROVED** [review]
LGTM and thanks for the update.

**Comment by shizhMSFT** [comment]
@ktarplee LoL. The DCO check failed. Could you fix that?

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**